src/libstd/sys_common/backtrace.rs

```diff
@@ -88,43 +90,84 @@ fn _print(w: &mut Write, format: PrintFormat) -> io::Result<()> {
         if has_more_filenames {
             w.write_all(b" <... and possibly more>")?;
         }
+        if cfg!(not(windows)) {
+            write!(w, "\x1B[0m")?;
+        }
     }
 
     Ok(())
 }
 
-/// Returns a number of frames to remove at the beginning and at the end of the
-/// backtrace, according to the backtrace format.
-fn filter_frames(frames: &[Frame],
-                 format: PrintFormat,
-                 context: &BacktraceContext) -> (usize, usize)
-{
-    if format == PrintFormat::Full {
-        return (0, 0);
-    }
+fn should_show_frame(frame: &Frame, context: &BacktraceContext) -> bool {
+    const FILTERED_SYMBOLS: &[&str] = &[
+        "main",
+        "rust_begin_unwind",
+    ];
+    const FILTERED_SYMBOL_PARTS: &[&str] = &[
+        "panic",
+        "sys",
+        "lang_start",
+    ];
+    let mut should_show = true;
+    let _ = resolve_symname(*frame, |symname| {
+        if let Some(mangled_symbol_name) = symname {
+            for filtered_symbol in FILTERED_SYMBOLS {
+                if mangled_symbol_name == *filtered_symbol {
+                    should_show = false;
+                    return Ok(());
+                }
+            }
+            for filtered_symbol_part in FILTERED_SYMBOL_PARTS {
+                if mangled_symbol_name.contains(filtered_symbol_part) {
+                    should_show = false;
+                    return Ok(());
+                }
+            }
+        }
+        Ok(())
+    }, context);
+    should_show
+}
 
-    let skipped_before = 0;
+/// Returns the frames to show as a Vec.
+/// If the bool is true the frame is on the edge between showing and not showing.
+fn filter_frames<'a>(frames: &'a [Frame],
+                     context: &BacktraceContext,
+                     format: PrintFormat) -> Vec<(usize, &'a Frame, bool)>
```

Comment:
Backtraces are currently generated in some sensitive contexts that often work best if we avoid allocation as much as possible, so could a callback or a different scheme be used instead of a vector?

Reply:
Yes, but it is harder, due to having to know if the previous or next frame have to be filtered out.

Reply:
Ok, I think this'll be a requirement for landing though